sdk/eventhub/azure-eventhub/azure/eventhub/_tracing.py

```diff
@@ -0,0 +1,270 @@
+        with span_impl_type(name="EventHubs.process", kind=SpanKind.CONSUMER, links=links) as span:
+            add_span_attributes(span, TraceOperationTypes.PROCESS, client, message_count=len(links))
+            yield
+    else:
+        yield
+
+
+def trace_message(
+    message: Union[uamqp_Message, Message],
+    amqp_transport: AmqpTransport,
+    additional_attributes: Optional[Dict[str, Any]] = None
+) -> Union[uamqp_Message, Message]:
+    """Add tracing information to the message and return the updated message.
+
+    Will open and close an message span, and add trace context to the app properties of the message.
+    """
+    try:
+        span_impl_type: Type[AbstractSpan] = settings.tracing_implementation()
+        if span_impl_type is not None:
+            with span_impl_type(name="EventHubs.message", kind=SpanKind.PRODUCER) as message_span:
+                headers = message_span.to_header()
+
+                if "traceparent" in headers:
+                    message = amqp_transport.update_message_app_properties(
+                        message,
+                        TRACE_DIAGNOSTIC_ID_PROPERTY,
+                        headers["traceparent"]
+                    )
+                    message = amqp_transport.update_message_app_properties(
+                        message,
+                        TRACE_PARENT_PROPERTY,
+                        headers["traceparent"]
+                    )
+                if "tracestate" in headers:
+                    message = amqp_transport.update_message_app_properties(
+                        message,
+                        TRACE_STATE_PROPERTY,
+                        headers["tracestate"]
+                    )
+
+                message_span.add_attribute(
+                    TraceAttributes.TRACE_NAMESPACE_ATTRIBUTE, TraceAttributes.TRACE_NAMESPACE)
+                message_span.add_attribute(
+                    TraceAttributes.TRACE_MESSAGING_SYSTEM_ATTRIBUTE, TraceAttributes.TRACE_MESSAGING_SYSTEM)
+
+                if additional_attributes:
+                    for key, value in additional_attributes.items():
+                        if value is not None:
+                            message_span.add_attribute(key, value)
+
+    except Exception as exp:  # pylint:disable=broad-except
+        _LOGGER.warning("trace_message had an exception %r", exp)
+
+    return message
+
+
+def get_span_links_from_received_events(events: Union[EventData, Iterable[EventData]]) -> List[Link]:
+    """Create span links from received events.
+
+    This will extract the traceparent and tracestate from the event properties and create span links
+    based on these values. The time the event was enqueued is also added as a link attribute.
+    """
+    # pylint:disable=isinstance-second-argument-not-valid-type
+    trace_events = events if isinstance(events, Iterable) else (events,)
+    links = []
+    try:
+        for event in trace_events:
+            if event.properties:
```

Comment:
would it be true if properties are empty? (excuse my python)

I believe we're adding on if traceparent/tracestate are missing (which is good).
Assuming properties are empty, we should still add a link if there are no properties at all - this way we'll (later on) be able to add attributes to links even when there is no context coming from producer and also calculate `messaging.batch.message_count` attribute properly - we're currently counting it as a number of links.

Reply:
Yea, empty properties equate to false. I can adjust it so that it always adds a link per event, even if the the context would be empty.

i.e, the links would look like:
```
        {
            "context": {
                "trace_id": "0x00000000000000000000000000000000",
                "span_id": "0x0000000000000000",
                "trace_state": "[]"
            },
            "attributes": {}
        }
```